### PR TITLE
adding new ThresholdPushbackPolicy

### LINKED ISSUE #442: Additional default PushBackPolicy
In `org.osgi.util.pushstream.PushbackPolicyOption` we have a couple of usefull default policies. Aver the last years, we found another one that became our default and I suggest adding it to the exsisting List:

```
	/**
	 * This {@link PushbackPolicy} increases linear after the queue breaches a given threshold.  
	 * If the threshold is e.g. 100 the 101st element would cause a backpressure of one and the 110th element a pushback of 10. 
	 */
	LINEAR_AFTER_THRESHOLD {
		@Override
		public <T, U extends BlockingQueue<PushEvent<? extends T>>> PushbackPolicy<T, U> getPolicy(long value) {
			return q -> Math.max(0, q.size() - value);
		}
	};
```

### PR DESCRIPTION
**Pull request by juergen-albert**
closes #442

Signed-off-by: Juergen Albert <j.albert@data-in-motion.biz>

### CODE REVIEW

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/package-info.java
Comment (bjhargrave) [inline org.osgi.util.pushstream/src/org/osgi/util/pushstream/package-info.java:36]: You also need to update the comment above to use 1.1 instead of 1.0.

File: osgi.specs/docbook/706/util.pushstream.xml
Comment (bjhargrave) [inline osgi.specs/docbook/706/util.pushstream.xml:1085]:
```suggestion
        <para>The <xref
        linkend="org.osgi.util.pushstream.ThresholdPushbackPolicy"
        xrefstyle="hyperlink"/> provides a specific implementation of a push
        policy. It covers a common case were back pressure needs to be applied 
        after a certain threshold in the buffer is exceeded. The initial pressure 
        and subsequent increase can be configured as well.</para>
```

File: osgi.specs/docbook/706/util.pushstream.xml
Comment (bjhargrave) [inline osgi.specs/docbook/706/util.pushstream.xml:1571]:
```suggestion
        <para>Added new push back policy implementation <xref
        linkend="org.osgi.util.pushstream.ThresholdPushbackPolicy"
        xrefstyle="hyperlink"/>.</para>
```

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (bjhargrave) [inline org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java:37]:
```suggestion
public final class ThresholdPushbackPolicy<T, U extends BlockingQueue<PushEvent< ? extends T>>>
```

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (bjhargrave) [inline org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java:81]:
```suggestion
		if (threshold < 0 || initial < 0L || step < 0L) {
```

We should use proper long constants.

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (bjhargrave) [inline org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java:54]:
```suggestion
		this(threshold, 1L, 1L);
```

Proper long constants.

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (bjhargrave) [inline org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java:86]:
```suggestion
			throw new IllegalArgumentException(String.format(
					"threshold, initial and step must each be larger than zero. Current values are threshold [%s], initial [%s], step [%s]",
					Integer.valueOf(threshold), Long.valueOf(initial),
					Long.valueOf(step)));
		}
```

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (bjhargrave) [inline org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java:53]: I thought we agreed to use static methods rather than public constructors.

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (bjhargrave) [inline org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java:96]:
```suggestion
			return 0L;
```

**Review by bjhargrave — CHANGES_REQUESTED** [review]

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/package-info.java
Comment (juergen-albert) [reply org.osgi.util.pushstream/src/org/osgi/util/pushstream/package-info.java:36]: done

File: osgi.specs/docbook/706/util.pushstream.xml
Comment (juergen-albert) [reply osgi.specs/docbook/706/util.pushstream.xml]: done

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (juergen-albert) [reply org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java]: done

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (juergen-albert) [reply org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java]: done

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (juergen-albert) [reply org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java]: done

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (juergen-albert) [reply org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java]: done

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (juergen-albert) [reply org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java]: Done. I hope I picked some sensible names for the different methods

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (juergen-albert) [reply org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java]: done

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (juergen-albert) [reply org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java]: thanks, used your version

File: osgi.specs/docbook/706/util.pushstream.xml
Comment (juergen-albert) [reply osgi.specs/docbook/706/util.pushstream.xml]: thanks, used your version

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (bjhargrave) [inline org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java:43]:
```suggestion
 * @since 1.1
 */
```

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (bjhargrave) [inline org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java:114]: The code requires non-negative values, but the exception messages says you must use positive values. This needs to be reconciled. My guess is you really want positive values, so the code test must be `<= 0'`.

The new TCK tests should also confirm that 0 is illegal to match this code.

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (bjhargrave) [inline org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java:86]: Need comment before param.

**Review by bjhargrave — CHANGES_REQUESTED** [review]

File: org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java
Comment (juergen-albert) [reply org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java:114]: 0 for any or all of its values is actually legal and having one or more of them as such might even make sense in some situation. Thus I allowed for it. One might argue that e.g. all initialized with 0 does not make sense, as it would effectively do nothing, but it is still legal and work fine.

Comment (timothyjward) [reply org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java:118]: > The new TCK tests should also confirm that 0 is illegal to match this code.

This should not be the case, at least for `initial` and `increment`. The JavaDoc for this class outlines uses `0` for both of these (not simultaneously) to get useful behaviours. Specifically a zero increment gives a “threshold fixed” policy, and a zero initial gives a “threshold linear” policy. 

As Jürgen points out, zero for both initial and increment is a bit strange but not actively broken. Zero for threshold collapses to a combination of linear plus fixed, and might be considered odd enough to be broken.

Comment (bjhargrave) [reply org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java:118]: OK, then the exception message is wrong since it states the values "must each be larger than zero".

Comment (juergen-albert) [reply org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java]: That is correct. Changed it accordingly.

**Review by bjhargrave — APPROVED** [review]

**Comment by bjhargrave** [comment]
OK, but the TCK for org.osgi.service.log.stream failed. Did you diagnose that?

**Comment by juergen-albert** [comment]
Sorry, I didn't catch that. I came to the same conclusion as you did, regarding the log implementation. How do we want to precede? I guess we wait until the PR is merged and released and then merge again?

**Comment by bjhargrave** [comment]
This PR is done. We just need to change to the updated Eclipse Log Stream impl when it is available. Until then, the Log Stream TCK is broken.

**Pull request merged**
